riscv/triggers.cc

```diff
@@ -52,6 +58,30 @@ void trigger_t::tdata3_write(processor_t * const proc, const reg_t val) noexcept
   sselect = (sselect_t)((proc->extension_enabled_const('S') && get_field(val, CSR_TEXTRA_SSELECT(xlen)) <= SSELECT_MAXVAL) ? get_field(val, CSR_TEXTRA_SSELECT(xlen)) : SSELECT_IGNORE);
 }
 
+bool trigger_t::textra_match(processor_t * const proc) const noexcept
+{
+  auto xlen = proc->get_xlen();
+  auto hsxlen = proc->get_xlen(); // use xlen since no hsxlen
+  state_t * const state = proc->get_state();
+
+  assert(sselect <= SSELECT_MAXVAL);
+  if (sselect == SSELECT_ASID) {
+    reg_t satp = state->satp->read();
+    reg_t asid = get_field(satp,  SATP_ASID(xlen));
+    if ((asid ^ svalue) & ((1 << ASIDMAX(xlen)) - 1))
+      return false;
+  }
+
+  mhselect_mode_t mode = mhselect_mode(mhselect, proc->extension_enabled('H'));
+  if (mode == MHSELECT_VMID_SELECT) { // 2 and 6 are vmid_select
+    reg_t vmid = get_field(state->hgatp->read(), HGATP_VMID(xlen));
+    if ((vmid ^ mhselect_compare()) & ((1 << VMIDMAX(hsxlen)) - 1))
```

Comment:
Using xor obfuscates what's happening here. The spec says:

> This trigger will only match if VMID in hgatp equals the lower VMIDMAX (defined in the Privileged Spec) bits of {mhvalue, mhselect[2]}.

The code should read as closely to the prose as possible:
```
if (vmid & ((1 << VMIDMAX(hsxlen)) - 1) == mhselect_compare()) ...
```